server/sonar-process-monitor/src/test/java/org/sonar/application/SchedulerImplTest.java

```diff
@@ -0,0 +1,384 @@
+    assertThat(processLauncher.commands).containsExactly(ES_COMMAND, WEB_LEADER_COMMAND, CE_COMMAND);
+
+    // all processes are up
+    processLauncher.processes.values().forEach(p -> assertThat(p.isAlive()).isTrue());
+
+    // processes are stopped in reverse order of startup
+    underTest.terminate();
+    assertThat(orderedStops).containsExactly(COMPUTE_ENGINE, WEB_SERVER, ELASTICSEARCH);
+    processLauncher.processes.values().forEach(p -> assertThat(p.isAlive()).isFalse());
+
+    // does nothing because scheduler is already terminated
+    underTest.awaitTermination();
+  }
+
+  private void enableAllProcesses() {
+    settings.set(ProcessProperties.CLUSTER_ENABLED, "true");
+  }
+
+  @Test
+  public void all_processes_are_stopped_if_one_process_goes_down() throws Exception {
+    Scheduler underTest = startAll();
+
+    processLauncher.waitForProcess(WEB_SERVER).destroyForcibly();
+
+    underTest.awaitTermination();
+    assertThat(orderedStops).containsExactly(WEB_SERVER, COMPUTE_ENGINE, ELASTICSEARCH);
+    processLauncher.processes.values().forEach(p -> assertThat(p.isAlive()).isFalse());
+
+    // following does nothing
+    underTest.terminate();
+    underTest.awaitTermination();
+  }
+
+  @Test
+  public void all_processes_are_stopped_if_one_process_fails_to_start() throws Exception {
+    enableAllProcesses();
+    SchedulerImpl underTest = newScheduler();
+    processLauncher.makeStartupFail = COMPUTE_ENGINE;
+
+    underTest.schedule();
+
+    processLauncher.waitForProcess(ELASTICSEARCH).operational = true;
+    processLauncher.waitForProcess(WEB_SERVER).operational = true;
+
+    underTest.awaitTermination();
+    assertThat(orderedStops).containsExactly(WEB_SERVER, ELASTICSEARCH);
+    processLauncher.processes.values().forEach(p -> assertThat(p.isAlive()).isFalse());
+  }
+
+  @Test
+  public void terminate_can_be_called_multiple_times() throws Exception {
+    Scheduler underTest = startAll();
+
+    underTest.terminate();
+    processLauncher.processes.values().forEach(p -> assertThat(p.isAlive()).isFalse());
+
+    // does nothing
+    underTest.terminate();
+  }
+
+  @Test
+  public void awaitTermination_blocks_until_all_processes_are_stopped() throws Exception {
+    Scheduler underTest = startAll();
+
+    Thread awaitingTermination = new Thread(() -> underTest.awaitTermination());
+    awaitingTermination.start();
+    assertThat(awaitingTermination.isAlive()).isTrue();
+
+    underTest.terminate();
+    // the thread is being stopped
+    awaitingTermination.join();
+    assertThat(awaitingTermination.isAlive()).isFalse();
+  }
+
+  @Test
+  public void restart_reloads_java_commands_and_restarts_all_processes() throws Exception {
+    Scheduler underTest = startAll();
+
+    processLauncher.waitForProcess(WEB_SERVER).askedForRestart = true;
+
+    // waiting for all processes to be stopped
+    boolean stopped = false;
+    while (!stopped) {
+      stopped = orderedStops.size() == 3;
+      Thread.sleep(1L);
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR') Remove this use of "Thread.sleep()". [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS2925)